File: src/data_utils.py
from __future__ import annotations

from collections import defaultdict
from datetime import date, timedelta
from typing import Any, Dict, List

LEVELS = ["a1", "a2", "b1", "b2", "c1", "c2"]
LEVEL_BY_TEST = {
    1: "a1",
    2: "a2",
    3: "b1",
    4: "b1",
    5: "b2",
    6: "b2",
    7: "c1",
    8: "c1",
    9: "c2",
    10: "c2",
}
FALLBACK_LEVEL_ORDER = {
    "listening": ["a1", "b1", "c1"],
    "reading": ["a2", "b2", "c2"],
}


def get_daily_word_counts(words: List[Dict[str, Any]]) -> Dict[str, int]:
    counts: Dict[str, int] = defaultdict(int)
    for word in words:
        date_value = word.get("dateAdded")
        if date_value:
            counts[date_value] += 1
    return dict(sorted(counts.items()))


def _get_test_level(test_number: Any, source_type: str, occurrence_index: int) -> str:
    if test_number is not None:
        try:
            level = LEVEL_BY_TEST.get(int(test_number), "a1")
            return level if level in LEVELS else "a1"
        except (TypeError, ValueError):
            pass

    fallback_levels = FALLBACK_LEVEL_ORDER.get(source_type, ["a1"])
    return fallback_levels[occurrence_index % len(fallback_levels)]


def summarize_question_breakdown(words: List[Dict[str, Any]]) -> Dict[str, Dict[str, int]]:
    summary = {"listening": {level: 0 for level in LEVELS}, "reading": {level: 0 for level in LEVELS}}
    occurrence_counts = {"listening": 0, "reading": 0}

    for word in words:
        source = word.get("source", {})
        source_type = source.get("type")
        if source_type not in summary:
            continue

        test_number = source.get("test")
        level = _get_test_level(test_number, source_type, occurrence_counts[source_type])
        summary[source_type][level] += 1
        occurrence_counts[source_type] += 1

    return summary


def build_heat_map_data(counts: Dict[str, int], end_date_text: str) -> List[Dict[str, Any]]:
    try:
        end_date = date.fromisoformat(end_date_text)
    except ValueError as exc:
        raise ValueError(f"Invalid ISO date: {end_date_text}") from exc

    start_date = end_date - timedelta(days=364)
    data: List[Dict[str, Any]] = []
    current = start_date

    while current <= end_date:
        iso_date = current.isoformat()
        data.append({
            "date": iso_date,
            "count": counts.get(iso_date, 0),
        })
        current += timedelta(days=1)

    return data


def build_level_trend_data(daily_stats: List[Dict[str, Any]], skill: str) -> List[Dict[str, Any]]:
    if not isinstance(daily_stats, list) or not skill:
        return []

    return [
        {
            "level": level,
            "color": {
                "a1": "#60a5fa",
                "a2": "#34d399",
                "b1": "#fbbf24",
                "b2": "#f472b6",
                "c1": "#a78bfa",
                "c2": "#fb7185",
            }[level],
            "points": [
                {"date": entry.get("date"), "value": int(entry.get(skill, {}).get(level, 0))}
                for entry in daily_stats
            ],
        }
        for level in LEVELS
    ]

File: src/data-utils.js
export const LEVELS = ['a1', 'a2', 'b1', 'b2', 'c1', 'c2'];

const LEVEL_BY_TEST = {
  1: 'a1',
  2: 'a2',
  3: 'b1',
  4: 'b1',
  5: 'b2',
  6: 'b2',
  7: 'c1',
  8: 'c1',
  9: 'c2',
  10: 'c2',
};

const FALLBACK_LEVEL_ORDER = {
  listening: ['a1', 'b1', 'c1'],
  reading: ['a2', 'b2', 'c2'],
};

export function getDailyWordCounts(words) {
  const counts = {};

  words.forEach((word) => {
    const dateValue = word.dateAdded;
    if (!dateValue) return;
    counts[dateValue] = (counts[dateValue] ?? 0) + 1;
  });

  return Object.fromEntries(Object.entries(counts).sort(([left], [right]) => left.localeCompare(right)));
}

export function inferLevelFromTest(testNumber, sourceType, occurrenceIndex) {
  if (testNumber !== undefined && testNumber !== null && !Number.isNaN(Number(testNumber))) {
    const normalized = Number(testNumber);
    return LEVEL_BY_TEST[normalized] ?? 'a1';
  }

  const fallbackLevels = FALLBACK_LEVEL_ORDER[sourceType] ?? ['a1'];
  return fallbackLevels[occurrenceIndex % fallbackLevels.length];
}

export function summarizeQuestionBreakdown(words) {
  const summary = {
    listening: { a1: 0, a2: 0, b1: 0, b2: 0, c1: 0, c2: 0 },
    reading: { a1: 0, a2: 0, b1: 0, b2: 0, c1: 0, c2: 0 },
  };
  const occurrenceCounts = { listening: 0, reading: 0 };

  words.forEach((word) => {
    const source = word.source ?? {};
    const type = source.type;
    if (!summary[type]) return;

    const level = inferLevelFromTest(source.test, type, occurrenceCounts[type]);
    summary[type][level] += 1;
    occurrenceCounts[type] += 1;
  });

  return summary;
}

export function buildHeatMapData(counts, endDateText) {
  const endDate = new Date(`${endDateText}T12:00:00Z`);
  const startDate = new Date(endDate);
  startDate.setDate(startDate.getDate() - 364);

  const result = [];
  const current = new Date(startDate);

  while (current <= endDate) {
    const isoDate = current.toISOString().slice(0, 10);
    result.push({
      date: isoDate,
      count: counts[isoDate] ?? 0,
    });
    current.setDate(current.getDate() + 1);
  }

  return result;
}

export function getHeatMapColor(count) {
  if (count === 0) return 'heatmap-empty';
  if (count >= 1 && count <= 5) return 'heatmap-level-1';
  if (count >= 6 && count <= 10) return 'heatmap-level-2';
  return 'heatmap-level-3';
}
